File: parsers/hlb.py
from __future__ import annotations

import re
from datetime import datetime

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
from categorize import Transaction

try:
    import pdfplumber
except ImportError:
    raise ImportError("pdfplumber is required: pip install pdfplumber")

# Column x-coordinate ranges derived from HLB Pay&Save PDF format
_DATE_X    = (25,  82)
_DESC_X    = (82,  295)
_DEP_X     = (350, 430)
_WD_X      = (430, 520)

_DATE_RE   = re.compile(r'^\d{2}/\d{2}/\d{4}$')
_AMOUNT_RE = re.compile(r'^\d[\d,]*\.\d{2}$')

# Words that begin the statement summary section on the last page — stop collecting
# description text when any line starts with one of these.
_DESC_STOP = {"Balance", "Pengeluaran", "Simpanan", "Notis", "Penting"}


def _in(x: float, col: tuple) -> bool:
    return col[0] <= x < col[1]


def _parse_amount(text: str) -> float:
    return float(text.replace(',', ''))


def parse(filepath: str, statement_year: int, statement_month: int) -> list[Transaction]:
    raw: list[dict] = []   # accumulated across all pages

    with pdfplumber.open(filepath) as pdf:
        for page in pdf.pages:
            words = sorted(page.extract_words(), key=lambda w: (w['top'], w['x0']))

            # page_txs holds only this page's transactions so header words at the top
            # of the page don't leak into the last transaction from the previous page
            page_txs: list[dict] = []

            page_height = page.height
            for w in words:
                x0, top, text = w['x0'], w['top'], w['text']

                # Skip footer content (bottom 8% of page contains HLB's registration text)
                if top > page_height * 0.92:
                    continue

                if _in(x0, _DATE_X) and _DATE_RE.match(text):
                    page_txs.append({
                        'date_str':   text,
                        'date_top':   top,
                        'desc':       [],
                        'withdrawal': None,
                        'deposit':    None,
                    })

                elif page_txs:
                    tx = page_txs[-1]
                    if _in(x0, _DESC_X):
                        tx['desc'].append((top, x0, text))
                    elif _in(x0, _WD_X) and _AMOUNT_RE.match(text) and tx['withdrawal'] is None:
                        tx['withdrawal'] = _parse_amount(text)
                    elif _in(x0, _DEP_X) and _AMOUNT_RE.match(text) and tx['deposit'] is None:
                        tx['deposit'] = _parse_amount(text)

            raw.extend(page_txs)

    transactions: list[Transaction] = []
    for tx in raw:
        if tx['withdrawal'] is None:
            continue  # income / deposit row, skip

        # Build description: group words by line (same top), join lines with " / "
        # A gap > 20px between lines means we've left the transaction area (e.g. the
        # statement summary section on the last page falls in the same x-range).
        parts = sorted(tx['desc'], key=lambda p: (p[0], p[1]))
        lines: list[str] = []
        cur_top: float | None = None
        cur_line: list[str] = []
        for top, _, word in parts:
            if cur_top is None or abs(top - cur_top) > 3:
                if cur_line:
                    lines.append(' '.join(cur_line))
                if cur_top is not None and (top - cur_top) > 20:
                    break  # section gap — stop here
                if word in _DESC_STOP:
                    break  # statement summary keyword — stop here
                cur_line = [word]
                cur_top = top
            else:
                cur_line.append(word)
        if cur_line:
            lines.append(' '.join(cur_line))

        description = ' / '.join(lines)

        date = datetime.strptime(tx['date_str'], '%d/%m/%Y').strftime('%Y-%m-%d')

        transactions.append(Transaction(
            date=date,
            description=description,
            amount=tx['withdrawal'],
            currency='MYR',
            payment_type='DuitNow QR / POS / Bank transfer',
            source_account='HLB Pay & Save',
        ))

    return transactions

File: webapp/src/parsers/hlb.js
import { loadPdf } from '../pdf/pdfWords.js';
import { makeTransaction } from '../categorize.js';

// Column x-coordinate ranges derived from the HLB Pay & Save PDF format
const DATE_X = [25, 82];
const DESC_X = [82, 295];
const DEP_X = [350, 430];
const WD_X = [430, 520];

const DATE_RE = /^\d{2}\/\d{2}\/\d{4}$/;
const AMOUNT_RE = /^\d[\d,]*\.\d{2}$/;

// Words that begin the statement summary section on the last page — stop
// collecting description text when any line starts with one of these.
const DESC_STOP = new Set(['Balance', 'Pengeluaran', 'Simpanan', 'Notis', 'Penting']);

// pdf.js keeps real space characters inside a single text item (pdfplumber's
// word-merging drops them) — stripping whitespace before classification keeps
// these regexes identical to the pdfplumber-based Python parser's. See
// src/parsers/rhbPdf.js for the empirical basis (no sample HLB PDF exists to
// verify against directly — this parser is ported defensively on the same
// assumption pending a real test file).
function norm(text) {
  return text.replace(/\s+/g, '');
}

function inRange(x, [lo, hi]) {
  return x >= lo && x < hi;
}

function parseAmount(text) {
  return parseFloat(text.replace(/,/g, ''));
}

/**
 * @param {File} file
 * @returns {Promise<import('../categorize.js').Transaction[]>}
 */
export async function parse(file) {
  const pdf = await loadPdf(file);
  const raw = []; // accumulated across all pages

  for (let pageNum = 1; pageNum <= pdf.numPages; pageNum++) {
    const words = await pdf.words(pageNum);
    const pageHeight = await pdf.pageHeight(pageNum);

    // pageTxs holds only this page's transactions so header words at the top
    // of the page don't leak into the last transaction from the previous page
    const pageTxs = [];

    for (const w of words) {
      const { x0, top, text } = w;

      // Skip footer content (bottom 8% of page contains HLB's registration text)
      if (top > pageHeight * 0.92) continue;

      const n = norm(text);
      if (inRange(x0, DATE_X) && DATE_RE.test(n)) {
        pageTxs.push({ dateStr: n, desc: [], withdrawal: null, deposit: null });
        continue;
      }

      if (!pageTxs.length) continue;
      const tx = pageTxs[pageTxs.length - 1];

      if (inRange(x0, DESC_X)) {
        tx.desc.push({ top, x0, text });
      } else if (inRange(x0, WD_X) && AMOUNT_RE.test(n) && tx.withdrawal === null) {
        tx.withdrawal = parseAmount(n);
      } else if (inRange(x0, DEP_X) && AMOUNT_RE.test(n) && tx.deposit === null) {
        tx.deposit = parseAmount(n);
      }
    }

    raw.push(...pageTxs);
  }

  const transactions = [];
  for (const tx of raw) {
    if (tx.withdrawal === null) continue; // income / deposit row, skip

    // Build description: group words by line (same top), join lines with " / ".
    // A gap > 20px between lines means we've left the transaction area (e.g. the
    // statement summary section on the last page falls in the same x-range).
    const parts = [...tx.desc].sort((a, b) => a.top - b.top || a.x0 - b.x0);
    const lines = [];
    let curTop = null;
    let curLine = [];
    // Mirrors parsers/hlb.py's loop exactly, including its final flush running
    // unconditionally after the loop (break or not) — ported for parity, not
    // independently re-verified against a real HLB PDF (none is available here).
    for (const { top, text: word } of parts) {
      if (curTop === null || Math.abs(top - curTop) > 3) {
        if (curLine.length) lines.push(curLine.join(' '));
        if (curTop !== null && top - curTop > 20) break;
        if (DESC_STOP.has(word)) break;
        curLine = [word];
        curTop = top;
      } else {
        curLine.push(word);
      }
    }
    if (curLine.length) lines.push(curLine.join(' '));

    const description = lines.join(' / ');

    const [d, m, y] = tx.dateStr.split('/');
    const date = `${y}-${m}-${d}`;

    transactions.push(makeTransaction({
      date,
      description,
      amount: tx.withdrawal,
      currency: 'MYR',
      payment_type: 'DuitNow QR / POS / Bank transfer',
      source_account: 'HLB Pay & Save',
    }));
  }

  return transactions;
}
